### PR TITLE
Revived Pattern: Number Series

### PR DESCRIPTION
**Pull request by JeremyVyska**
Revived from the archive, and with some useful added bits from having had to explain this system to several people in the past year.

### CODE REVIEW

**Review by JesperSchulz — APPROVED** [review]
Good read. We could maybe add a section on locking at some point. It's unfortunately an issue we see too often, which has severe performance implications on an entire environment. Or maybe that's a guideline of its own.

**Comment by JeremyVyska** [comment]
Agreed.  Call this a v1 and maybe I can crosscheck against Kennie's guides for v2  😄

File: content/docs/patterns/no-series/index.md [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+```AL  
+field(107; "No. Series"; Code[20])
+{
+    Caption = 'No. Series';
+    Editable = false;
+    TableRelation = "No. Series";
+}
+```
+{{% alert title="Note" color="warning" %}}
+The **TableRelation** is important, and the **Editable** being false is advised.
```
Comment (TheDoubleH) [inline content/docs/patterns/no-series/index.md:65]:
```suggestion
The **`TableRelation`** is important, and the **`Editable`** being false is advised.
```

File: content/docs/patterns/no-series/index.md [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+One additional (and somewhat optional) feature in the Number Series engine allows multiple sequences per type, called **Relationships**.  For example, different numbers for Items that are finished goods versus raw materials.  This requires additional hooks on the Page.
+{{% /alert %}}
+
+## Usage in Business Central
+
+To understand an example use in the Base App, the Customer data entity is a good choice.  
+
+Implementation to connect the Customer "No." field to the Number Series engine is done at the table level.  The Customer table contains:
+
+A field to contain the number (typically the primary key), which will be Code type, length of 20:
```
Comment (TheDoubleH) [inline content/docs/patterns/no-series/index.md:40]: Has there even been a scenario, where the No. Series does not populate the Primary Key?

```suggestion
A field to contain the number (typically the primary key), which will be of type **`Code`**, length of **20**:
```

File: content/docs/patterns/no-series/index.md [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    Caption = 'No. Series';
+    Editable = false;
+    TableRelation = "No. Series";
+}
+```
+{{% alert title="Note" color="warning" %}}
+The **TableRelation** is important, and the **Editable** being false is advised.
+{{% /alert %}}
+
+And on the **OnInsert** trigger, code populates the **No. Series** and **No.** field.
```
Comment (TheDoubleH) [inline content/docs/patterns/no-series/index.md:68]:
```suggestion
And on the **`OnInsert`** trigger, code populates the **`No. Series`** and **`No.`** field.
```

File: content/docs/patterns/no-series/index.md [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        SalesSetup.Get();
+        SalesSetup.TestField("Customer Nos.");
+        if NoSeriesMgt.SelectSeries(SalesSetup."Customer Nos.", OldCust."No. Series", "No. Series") then begin
+            NoSeriesMgt.SetSeries("No.");
+            Rec := Cust;
+            OnAssistEditOnBeforeExit(Cust);
+            exit(true);
+        end;
+    end;
+end;
```
Comment (TheDoubleH) [inline content/docs/patterns/no-series/index.md:144]: `WITH` is being depreciated. Should we really keep this in a newly described pattern?

I know that is from the base app, but still?

File: content/docs/patterns/no-series/index.md [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+The function call takes the following parameters:
+
+```AL
+procedure InitSeries(
+  DefaultNoSeriesCode: Code[20]; 
+  OldNoSeriesCode: Code[20];
+  NewDate: Date;
+  var NewNo: Code[20];
+  var NewNoSeriesCode: Code[20])
+```
```
Comment (TheDoubleH) [inline content/docs/patterns/no-series/index.md:107]:
```suggestion
```AL
procedure InitSeries(
    DefaultNoSeriesCode: Code[20]; 
    OldNoSeriesCode: Code[20];
    NewDate: Date;
    var NewNo: Code[20];
    var NewNoSeriesCode: Code[20])
```
```

File: content/docs/patterns/no-series/index.md [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+A field to contain the number (typically the primary key), which will be Code type, length of 20:
+
+```AL
+field(1; "No."; Code[20])
+{
+    Caption = 'No.';
+
+    trigger OnValidate()
+    begin
+        [...]
```
Comment (TheDoubleH) [inline content/docs/patterns/no-series/index.md:49]: Should we add the code that validates that a manual number is OK?

**Review by TheDoubleH — CHANGES_REQUESTED** [review]
- Some formatting suggestions.
- Questioning if doing `with` is really appropriate
- requesting code sample for validating manual numbers

File: content/docs/patterns/no-series/index.md (hunk 2 above)
Comment (JeremyVyska) [reply content/docs/patterns/no-series/index.md:40]: Gen Jnl.

File: content/docs/patterns/no-series/index.md (hunk 4 above)
Comment (JeremyVyska) [reply content/docs/patterns/no-series/index.md:144]: That one should get a warning tag, that's true.  the WITH is in the Base App, but probably a good idea to call it out as a legacy issue that should not be repeated.

File: content/docs/patterns/no-series/index.md (hunk 6 above)
Comment (JeremyVyska) [reply content/docs/patterns/no-series/index.md:49]: Added

File: content/docs/patterns/no-series/index.md (hunk 2 above)
Comment (JeremyVyska) [reply content/docs/patterns/no-series/index.md]: Added a whole new section for it, since it's a different mechanism.  Good call.

**Review by TheDoubleH — APPROVED** [review]

**Review by waldo1001 — APPROVED** [review]

**Pull request merged**